v1.0.0-rc.1 (#2)
ucan-wg/delegation: README.md

# 0. Abstract

This specification describes the semantics and serialization format for [UCAN] delegation between principals. UCAN Delegation provides a cryptographically verifiable container, batched capabilities, hierarchical authority, and extensible caveats.
  
# 1. Introduction

UCAN Delegation is a certificate capability system with runtime-extensibility, ad hoc caveats, cacheability, and focused on ease of use and interoperability. Delegations act as a proofs for [UCAN Invocation]s.

Delegation provides a way to "transfer authority without transferring cryptographic keys". As an authorization system, it is more interested in "what" can be done than a list of "who can do what". For more on how Delegation fits into UCAN, please refer to the [high level spec][UCAN].

# 2. Delegation (Envelope)

Delegations MUST include a signature that validates against the `iss` DID. A Delegation Payload on its own MUST NOT be considered a valid Delegation.

| Field   | Type        | Required | Description                                  |
|---------|-------------|----------|----------------------------------------------|

> | `ucd`   | `&Payload`  | Yes      | The CID of the [Delegation Payload][Payload] |

Author: I don't know if we need to enforce it, but I expect to use inline IPLD a _lot_

> | `ucd`   | `&Payload`  | Yes      | The CID of the [Delegation Payload][Payload] |

Author: Can you say more about why this is a problem in js-ipld? That seems like a pretty serious limitation in such a library. Isn't the block limit in IPFS (not IPLD) something like 2MB?

> | `ucd`   | `&Payload`  | Yes      | The CID of the [Delegation Payload][Payload] |

Reviewer: Yes recommended limit is around 2MB. 

> Can you say more about why this is a problem in js-ipld?

Before I go into JS specifics here I think fundamental language agnostic issue is what if I don't have a linked block ? When it's inline field that issue simply does not exist.

As of JS specifically, everything is built around [js-multiformats](https://github.com/multiformats/js-multiformats) library which has notion of the block and codecs that can encode / decode blocks. There is no notion of codecs that span multiple blocks nor generic structure for traversing objects that span multiple blocks. There had been several attempts to address this but nothing sticked and I think that is because it introduces asynchrony which in JS is far more consequential and invasive than other languages.

In go stack I think folks have embraced IPLD Prime which takes care of block loading behind the scenes and they can also block execution if needed so it's less of the issue there.

In Rust async also not as big of deal you can always block and wait, so I think loading blocks seems fine there, although having `Result` is probably not great there either.

> | `sig`   | `Signature` | Yes      | The `iss`'s [Signature] over the `ucd` field |

Author: Yep

| `udv` | `String`                                  | Yes      | UCAN Semantic Version (`1.0.0-rc.1`)                        |
| `iss` | `DID`                                     | Yes      | Issuer DID (sender)                                         |
| `aud` | `DID`                                     | Yes      | Audience DID (receiver)                                     |
| `nbf` | `Integer` (53-bits[^js-num-size])         | No       | "Not before" UTC Unix Timestamp in seconds (valid from)     |
| `exp` | `Integer \| null` (53-bits[^js-num-size]) | Yes      | Expiration UTC Unix Timestamp in seconds (valid until)      |
| `nnc` | `String`                                  | Yes      | Nonce                                                       |
| `mta` | `{String : Any}`                          | No       | [Meta] (asserted, signed data) — is not delegated authority |
| `sub` | `DID`                                     | Yes      | Principal that the chain is about (the [Subject])           |
| `can` | `String`                                  | Yes      | [Ability]                                                   |

> | `iff` | `[Caveat]`                                | Yes      | [Caveat]s                                                   |

Author: Switched to `cond`

The payload MUST be serialized as [IPLD] and [signed over][Envelope]. The RECOMMENDED IPLD codec is [DAG-CBOR].

## 3.1 Version

> The `udv` field sets the version of the UCAN Delegation specification used in the payload.

Reviewer 2: i think i know what you mean, but can you give a quick example to illustrate what you'd like?

> The `udv` field sets the version of the UCAN Delegation specification used in the payload.

Reviewer: > i think i know what you mean, but can you give a quick example to illustrate what you'd like?

I was referring to what IPLD Schema calls [keyed unions](https://ipld.io/docs/schemas/using/authoring-guide/#keyed) which later got named "capsule types". What we have in this spec is:

```ts
{
   udv: "1.0.0",
   ...ucan
}
```

If your code can take ☝️ or some other type it becomes hard to guess which one you get. What I was suggesting instead is following:

```
{ "ucan@1.0.0": ucan }
```

Now you it's is clear that it is UCAN of specific version, better yet you can extend that type into a union with more versions you need to support and maybe even other things like invocations, revocations etc...

Downside is languages like TS or Rust aren't really fond of that pattern and you always have to wrap, while in previous pattern if you are aware what is it from context you don't have to wrap, but if you don't you could do something like `{ ucan: { udv: "1.0.0", ...ucan } }`

> The `udv` field sets the version of the UCAN Delegation specification used in the payload.

Reviewer 3: Rust is actually pretty happy with capsule types. We've switched to them in WNFS/rs-wnfs, and it's made some things easier.

However, putting a semantic version inside the key *would* be somewhat harder to deal with, since then you'd like to match on "ucan@1.0.x" or "ucan@1.x.x" in some cases. Putting the version inside is what we do in WNFS, but again that becomes less neat when you want have different fields for different versions. You *could* put the major version into the key, and the rest into the record, but that get's *very* unconventional fast: `{ "ucan@v1": { "subversion": "0.0", ... } }`.

## 3.2 Principals

The `iss` and `aud` fields describe the token's principals. They are distinguished by having DIDs. These can be conceptualized as the sender and receiver of a postal letter. The token MUST be signed with the private key associated with the DID in the `iss` field. Implementations MUST include the [`did:key`] method, and MAY be augmented with [additional DID methods][DID].

The `iss` and `aud` fields MUST contain a single principal each.

> If an issuer's DID has multiple or mutable keys (e.g. [`did:plc`], [`did:web`]), the key used to sign the UCAN MUST be made explicit, using the [DID fragment] (the hash index) in the `iss` field. The `aud` field SHOULD NOT include a hash field, as this defeats the purpose of delegating to an identifier for multiple keys instead of an identity.

Author: haha I like how we've changed positions on this since a much older version. I'm okay with that, sure. Will update 👍

> If an issuer's DID has multiple or mutable keys (e.g. [`did:plc`], [`did:web`]), the key used to sign the UCAN MUST be made explicit, using the [DID fragment] (the hash index) in the `iss` field. The `aud` field SHOULD NOT include a hash field, as this defeats the purpose of delegating to an identifier for multiple keys instead of an identity.

Reviewer 2: > The `aud` field SHOULD NOT include a hash field, as this defeats the purpose of delegating to an identifier for multiple keys instead of an identity.

> include a hash field

what does 'a hash field' mean here? (i.e. how would i implement/test this behavior)

 
[EdDSA] `did:key`s MUST be supported, and their use is RECOMMENDED. [RSA][did:key RSA] and [P-256 ECDSA][did:key ECDSA] `did:key`s MUST be supported, but SHOULD NOT be used when other options are available.

Note that every [Subject] MUST correspond to a root delegation issuer.

Below are a couple examples:

```json
"aud": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp",
"iss": "did:key:zDnaerDaTF5BXEavCrfRZEk316dpbLsfPDZ3WJ5hRTPFU2169",
```

```json
"aud": "did:web:example.com",
"iss": "did:pkh:eth:0xb9c5714089478a327f09197987f16f9e5d936e8a",
```

```json
"aud": "did:plc:ewvi7nxzyoun6zhxrhs64oiz",
"iss": "did:web:example.com",
```

```json
"aud": "did:pkh:eth:0xb9c5714089478a327f09197987f16f9e5d936e8a",

> "iss": "did:plc:ewvi7nxzyoun6zhxrhs64oiz",

Author: How so?

  "can": "crud/create",
  "iff": [
    {
      "uri": "https://example.com/blog/", // Resource
      "status": "draft"
    }
  ],
  // ...
}
```

## 4.3 Ability

Abilities MUST be presented as a case-insensitive string. By convention, abilities SHOULD be namespaced with a slash, such as `msg/send`. One or more abilities MUST be given for each resource.

``` js
{
  "sub": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp": {
  "can": "msg/send",
  "iff": [
    {
      "from": "mailto:alice@example.com",
      "to": "mailto:bob@example.com"
    }
  ],
  // ...
}
```

Abilities MAY be organized in a hierarchy that abstract over many [Command]s. A typical example is a superuser capability ("anything") on a file system. Another is mutation vs read access, such that in an HTTP context, `write` implies `put`, `patch`, `delete`, and so on. [`*` gives full access][Wildcard Ability] to a Resource more in the vein of object capabilities. Organizing abilities this way allows Resources to evolve over time in a backward-compatible manner, avoiding the need to reissue UCANs with new resource semantics.

### 4.3.1 Reserved Abilities

#### 4.3.1.1 `ucan` Namespace

> The `ucan` ability namespace MUST be reserved. This MUST include any ability string matching the regex `/^ucan.*/`

Author: Any concerns about nonprinting chars?

> The `ucan` ability namespace MUST be reserved. This MUST include any ability string matching the regex `/^ucan.*/`

Author: Yeah nevermind. Nonprinting can happen anywhere. Fixed.

``` js
[
  {       // ┐
    a: 1, // ├─ Caveat ─┐
    b: 2  // │          │
  },      // ┘          ├─ AND ─┐
  {       // ┐          │       │
    c: 3  // ├─ Caveat ─┘       │
  },      // ┘                  ├─ AND
  {       // ┐                  │
    d: 4  // ├─ Caveat ─────────┘
  }       // ┘
]
```

Expressing caveats in this standard way simplifies ad hoc extension at delegation time. As a concrete example, if a root UCAN caveat has a `tag` field but no (plural) `tags` field, it is still possible to express multiplicity by adding another `AND`ed caveat:

``` js
// Original Caveat
[
  {
    "uri": "https://blog.example.com/",
    "tag": "news"
  }
]

// Attenuated Caveat
[
  {
    "uri": "https://blog.example.com/",
    "tag": "news"
  }, 
  // AND
  {
    "tag": "breaking"
  }
]
```

In this example, the original caveat had not accounted for there being multiple tags at runtime. The attenuated capability grants access to blog posts at `https://blog.example.com/` that are tagged with both `news` and `breaking` due to the `AND` in the predicate.

> 

Reviewer 3: Personally, I'd love for caveats to simply be arbitrary metadata.
If a service starts out with a `{ "tag": "something" }` caveat, then it can eventually upgrade to `{ "tags": [...] }` by parsing the old format into the new.
I want libraries like rs-ucan to be dumb. They should not try to figure out how caveats subsume or not subsume each other nor should the UCAN spec define *the* way of how caveats subsume each other.

Caveat subsumption can be complex. It's not far-fetched to think that a use case may want to do complex cryptography, or perhaps (as in the ucanto RPC case) a full expression language on a level similar to SQL.
At the same time there will be use-cases that just don't need any of this.

As an implementer of a service, I wouldn't mind having to provide the exact caveat subsumption logic to rs-ucan (and perhaps being able to pick a simple default one that's similar to the one described in this spec).
However, I'd be unhappy if I'm forced to specify my caveats in exactly the format that UCAN delegations define, while a simpler service-specific definition would work.

The other side of the coin may be interoperability, but I don't see how yet.
We've always had "services describe their own semantics for delegation", and that's worked out? (Right?)

> 

Author: Yeah I agree with @matheus23. Historically UCAN has been heavy on semantics and light on syntactic rules. I'm increasingly convinced there where we can make things syntactic, the libraries can become a lot "dumber". There's still a plenty of semantics to contend with, but we can reduce that surface area significantly.

This is also helpful if each object has a special meaning or sense:

``` js
[
  {
    "type": "path",
    "segments": ["blog", "october"]
  },
  {
    "type": "market",
    "segments": ["manufacturing", "healthcare", "service", "technology"]
  }
]
```

Note that while adding whole objects is useful in many situations (as above), attenuation MAY also be achieved by adding fields to an object:

``` js
// Original Caveat
[
  {
    "uri": "https://blog.example.com/",
    "tag": "news"
  }
]

// Attenuated Caveat
[
  {
    "uri": "https://blog.example.com/",
    "tag": "news",
    "status": "draft" // New field
  }
]

> ```

Reviewer: Let me expand on my points around having proper operators as opposed to array of caveats that only executor will know how to merge / interpret. This is something I have been thinking for a long time and was planning to implement in user space specifically in ucanto RPC. Here is are couple of examples from our set of capabilities

```js
// delegate all store abilities without any constraints
{
   sub: "did:key:zSpace"
   can: "store/*",
}

// can re-delegate ☝️ by constraining to specific ability and limiting write size
{
   sub: "did:key:zSpace"
   can: "store/add",
   whr: {
      size: { "<=": 2222 }
   }
}

// can re-delegate as long as new constraints do not violate ☝️
{
   sub: "did:key:zSpace"
   can: "store/add",
   whr: {
      size: { "<=": 2000, ">": 1000 } // size is within given range
   }
}

// can re-delegate as long as new constraints do not violate ☝️
{
   sub: "did:key:zAlice",
   can: "store/add",
   whr: {
      archive: { "=": { "/": "bag...car" } },   // can only add specific CAR file
      size: { "=": 1234 }                               // which must have this exact size
   }
}
```

Here is the tags example from the above

```js
// Original Caveat
{
    "uri": { "=": "https://blog.example.com/" },
    "tag": { "⊇": ["news"] } // superset of the following tags
  }

// Attenuated Caveat (Must satisfy ☝️ caveats)
{
    "uri": { "=": "https://blog.example.com/" },
    "tag": { "⊇": ["news", "breaking"] } // superset of the following tags
  }

// Attenuated Cavet (Must satisfy ☝️ caveats)
{
    "uri": { "=": "https://blog.example.com/" },
    // can be super set of the specified tags but may not include other tags
    "tag": { "⊇": ["news", "breaking"], "⊄": ["blockchain"] }
}
```

We could also do something similar with arbitrary strings / paths

```js
// Delegates access to path as long as it starts with /public/ 
{
  "path": { "|=": "/public/" }
}

// Can attenuate as long as ☝️ caveat is satisfied
{
   "path": { "|=": "/public/photos/" } // as long as starts with /public/photos/
}

// Can attenuate as long as ☝️ caveat is satisfied
{
  "path": {
       // starts with /public/photos/
       "|=": "/public/photos/",
        // ends with ".jpeg"
        "=|": ".jpeg"
   } 
}

// Can attenuate as long as ☝️ caveat is satisfied
{
   "path": {
      "=": "/public/photos/profile.jpeg"
   }
}
```

None have OR operators even though in some case it may be handy, however that's ok as ORs can be achieved through separate delegations

> ```

Author: That's basically what we have, but we're not going to ship a bunch of UTF8. There will be a standard library to drive these mainly synatctically, but we need this to be extensible. WNFS has use cases that cannot be handled with the above, for example.

> ```

Author: Instead of this..

```json
// Original Caveat
{
    "uri": { "=": "https://blog.example.com/" },
    "tag": { "⊇": ["news"] } // superset of the following tags
  }

// Attenuated Caveat (Must satisfy ☝️ caveats)
{
    "uri": { "=": "https://blog.example.com/" },
    "tag": { "⊇": ["news", "breaking"] } // superset of the following tags
  }
```

...where you have to understand how to prove each caveat, we think it would be much easier to just have them as separate conditions. Using your notation:

```json
{
    {"field": "uri", "=": "https://blog.example.com/"},
    {"field": "tag", "include": "news" },
    {"field": "tag", "include": "breakling" }
  }
```

In this version, you can treat the array as an `all` predicate, and you don't have to understand the structure of each condition. At each delegation, it's just "did we add more conditions and leave the old ones as they were?"

> ```

Author: We also don't need `"="` in the caveats (now "conditions"), because that's what's precisely captured by the `args` field

Each capability has its own semantics, which needs to be interpretable by the [Executor]. Therefore, a validator MUST NOT reject all capabilities when one that is not relevant to them is not understood. For example, if a caveat fails a delegation check at execution time, but is not relevant to the invocation, it MUST be ignored.

If _any_ of the following criteria are not met, the UCAN MUST be considered invalid:

## 5.1 Time Bounds

A UCAN's time bounds MUST NOT be considered valid if the current system time is before the `nbf` field or after the `exp` field. This is called the "validity period." Proofs in a chain MAY have different validity periods, but MUST all be valid at execution-time. This has the effect of making a delegation chain valid between the latest `nbf` and earliest `exp`.

``` js
// Pseudocode

const ensureTime = (delegationChain, now) => {
  delegationChain.forEach((ucan) => {
    if (!!ucan.nbf && now < can.nbf) {
      throw new Error(`Delegation is not yet valid, but will become valid at ${ucan.nbf}`)
    }

    if (ucan.exp !== null && now > ucan.exp) {
      throw new Error(`Delegation expired at ${ucan.exp}`)
    }
  })
}

> ```

Author: Time bounds no longer need to get narrower. It's impractical in this updated version with late delegation bindings. We no longer have any cache benefit by enforcing that narrowing, so it only matters that they're valid at invocation time.

## 5.2 Principal Alignment

> In delegation, the `aud` field of every proof MUST match the `iss` field of the UCAN being delegated to. This alignment MUST form a chain back to the originating principal for each resource. 

Author: Good point!

        subgraph del1 [Delegated UCAN]
            del1Iss(iss: Bob) --> rootAud
            del1Sub(sub: Alice)
            del1Aud(aud: Carol)
            del1Cap("cap: (Storage, crud/*)") --> rootCap

            del1Sub --> rootSub
        end

        subgraph del2 [Delegated UCAN]
            del2Iss(iss: Carol) --> del1Aud
            del2Sub(sub: Alice)
            del2Aud(aud: Dan)
            del2Cap("cap: (Storage, crud/*)") --> del1Cap

            del2Sub --> del1Sub
        end
    end

     subgraph inv [Invocation]
        invIss(iss: Dan)
        args("args: [Storage, crud/update, (key, value)]")
        invSub(sub: Alice)
        prf("proofs")
    end

    invIss --> del2Aud
    invoker --> invIss
    args --> del2Cap
    invSub --> del2Sub
    rootIss --> subject
    rootSub --> subject
    prf --> Delegations
```

In the above diagram, Alice has some storage. This storage may exist in one location with a single source of truth, but to help build intuition this example is location independent: local versions and remote stored copies are eventually consistent, and there is no one "correct" copy. As such, we list the owner (Alice) directly on the resource.

> Alice delegates access to Bob. Bob then redelegates to Carol. Carol invokes the UCAN as part of a REST request to a compute service. To do this, she MUST both provide proof that she has access (the UCAN chain), and MUST delegate access to the invoking compute service. The invoking service MUST check that the root issuer (Alice) is in fact the owner (typically the creator) of the resource. This MAY be listed directly on the resource, as it is here. Once the UCAN chain and root ownership are validated, the storage service performs the write.

Author: Hmm yeah I... honestly don't know what I'm trying to say there. I _think_ I started showing a case that got moved to teh high level spec. I'll just delete the entire paragraph.

### 5.2.1 Recipient Validation

An agent executing a capability MUST verify that the outermost `aud` field _matches its own DID._ The associated ability MUST NOT be performed if they do not match. Recipient validation is REQUIRED to prevent the misuse of UCANs in an unintended context.

The following UCAN fragment would be valid to invoke as `did:key:zH3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV`. Any other agent MUST NOT accept this UCAN. For example, `did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp` MUST NOT run the ability associated with that capability.

``` js
{
  "aud": "did:key:zH3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV",
  "iss": "did:key:zAKJP3f7BD6W4iWEQ9jwndVTCBq8ua2Utt8EEjJ6Vxsf",
  // ...
}
```

A good litmus test for invocation validity by a invoking agent is to check if they would be able to create a valid delegation for that capability.

## 5.3 Caveat Attenuation

> The caveat array SHOULD NOT be empty, as an empty array means "in no case" (which is equivalent to not listing the ability). This follows from the rule that delegations MUST be of equal or lesser scope. When an array is given, an attenuated caveat MUST (syntactically) include all of the fields of the relevant proof caveat, plus the newly introduced caveats.

Author: Fixed / updated / simplied in latest version

Here are some abstract examples:

| Proof Caveats        | Delegated Caveats  | Is Valid? | Comment                               |
|----------------------|--------------------|-----------|---------------------------------------|
| `[]`                 | `[]`               | Yes       | Equal                                 |
| `[]`                 | `[{}]`             | Yes       | [True Caveat]                         |
| `[{}]`               | `[]`               | Yes       | [True Caveat]                         |
| `[{}]`               | `[{}]`             | Yes       | Equal                                 |
| `[{a: 1}]`           | `[{a: 1}]`         | Yes       | Equal                                 |
| `[{}]`               | `[{a: 1}]`         | Yes       | Attenuates `{}` by adding fields      |

> | `[{a: 1}], [{b: 2}]` | `[{a: 1, b: 2}]`   | Yes       | Attenuates existing caveat            |

Author: This is greatly simplified in the latest version

# UCAN Delegation Specification v1.0.0-rc.1

## Editors

- [Brooklyn Zelenka], [Fission]

## Authors

- [Brooklyn Zelenka], [Fission]
- [Daniel Holmgren], [Bluesky]
- [Irakli Gozalishvili], [Protocol Labs]
- [Philipp Krüger], [Fission]

## Dependencies

- [DID]
- [IPLD]
- [UCAN]

## Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119].

# 0. Abstract

This specification describes the semantics and serialization format for [UCAN] delegation between principals. UCAN Delegation provides a cryptographically verifiable container, batched capabilities, hierarchical authority, and extensible conditions.
  
# 1. Introduction

UCAN Delegation is a certificate capability system with runtime-extensibility, ad hoc conditions, cacheability, and focused on ease of use and interoperability. Delegations act as a proofs for [UCAN Invocation]s.

> Delegation provides a way to "transfer authority without transferring cryptographic keys". As an authorization system, it is more interested in "what" can be done than a list of "who can do what". For more on how Delegation fits into UCAN, please refer to the [high level spec][UCAN].

Reviewer 2: ```suggestion
Delegation provides a way to "transfer authority without transferring cryptographic keys". As an authorization system, it is more interested in "what can be done?" than a list of "who can do what?". For more on how Delegation fits into UCAN, please refer to the [high level spec][UCAN].
```

editorial/consistency suggestion but no worries if you dont like it

Author: Thanks @gobengo ! Had some issues with the automated merge, so added manually :)

| `exp`   | `Integer` (53-bits[^js-num-size]) | Yes      | Expiration UTC Unix Timestamp in seconds (valid until)      |
| `nonce` | `Bytes`                           | Yes      | Nonce                                                       |
| `meta`  | `{String : Any}`                  | No       | [Meta] (asserted, signed data) — is not delegated authority |
| `sub`   | `DID`                             | Yes      | Principal that the chain is about (the [Subject])           |
| `can`   | `String`                          | Yes      | The [Command] to eventually invoke                          |
| `args`  | `{String : Any}`                  | Yes      | Any [Arguments] that MUST be present in the Invocation      |
| `cond`  | `[Condition]`                     | Yes      | Any additional [Condition]s                                 |

The payload MUST be serialized as [IPLD] and [signed over][Envelope]. The RECOMMENDED IPLD codec is [DAG-CBOR].

## 3.1 Version

The `udv` field sets the version of the UCAN Delegation specification used in the payload.

## 3.2 Principals

> The `iss` and `aud` fields describe the token's principals. They are distinguished by having DIDs. These can be conceptualized as the sender and receiver of a postal letter. The token MUST be signed with the private key associated with the DID in the `iss` field. Implementations MUST include the [`did:key`] method, and MAY be augmented with [additional DID methods][DID].

Reviewer 2: ```suggestion
The `iss` and `aud` fields describe the token's principals. They are distinguished by having DIDs. These can be conceptualized as the sender and receiver of a postal letter. The token MUST be signed with a [verification method](https://www.w3.org/TR/did-core/#verification-methods) with [verification relationship](https://www.w3.org/TR/did-core/#verification-relationships) [`capabilityDelegation`](https://www.w3.org/TR/did-core/#capability-delegation) in the [DID document](https://www.w3.org/TR/did-core/#dfn-did-documents) that is the result of resolving the DID in the `iss` field of the UCAN. Implementations MUST include the [`did:key`] method, and MAY be augmented with [additional DID methods][DID].
```

Reviewer 4: maybe 
`from the DID in the `iss` field of the UCAN. Implementations MUST include the [`did:key`] method, and MAY be augmented with [additional DID methods][DID].` 
-->
`from the DID in the `iss` field of the UCAN. Implementations MUST include the [`did:key`] method, and MAY be augmented with [additional DID methods][DID] that support the above flow.`

Reviewer 2: > (Not all DID methods support the key names capabilityDelegation or arbitrary key names!)

I think it's fine with text as is. did-core defines capability delegation so that's all i need to refer to without worrying about methods that implemented it (fully/partially/correctly/incorrectly). If in practice someone is using UCANs with a did method that doesn't support capability delegations.... well IMO I think that should basically be inherently impossible but doesn't require spec text qualifications to point it out. did using method that doesn't support capability verification methods is just like using a did method that does, but the did doc has an empty array of them

Reviewer 4: ![image](https://github.com/ucan-wg/delegation/assets/37127325/ca58542c-42f9-4e74-affa-55ffcb5b6d3e)
very few of the registered DID methods implement this optional property, tho? IMHO everything that the DID method spec defines as optional is "rely on or assume support for at your own risk"...

  "args":{
    "uri": "https://example.com/blog/", // Resource
    "status": "draft"
  },
  // ...
}
```

## 4.3 Command

Commands MUST be presented as a case-insensitive string. By convention, Commands SHOULD be namespaced with a slash, such as `msg/send`.

``` js
{
  "sub": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp": {
  "can": "msg/send",
  "arg": {
    "from": "mailto:alice@example.com",
    "to": "mailto:bob@example.com"
  },
  // ...
}
```

Abilities MAY be organized in a hierarchy that abstract over many [Command]s. A typical example is a superuser capability ("anything") on a file system. Another is mutation vs read access, such that in an HTTP context, `write` implies `put`, `patch`, `delete`, and so on. [`*` gives full access][Wildcard Ability] to a Resource more in the vein of object capabilities. Organizing abilities this way allows Resources to evolve over time in a backward-compatible manner, avoiding the need to reissue UCANs with new resource semantics.

### 4.3.1 Reserved Commands

#### 4.3.1.1 `ucan/` Namespace

The `ucan/` ability namespace MUST be reserved. This MUST include any ability string matching the regex `/^ucan\/.*/`

> Support for the `ucan/*` delegated proof ability is RECOMMENDED.

Reviewer 3: This seems to be the only mention of the `ucan/*` ability. 
It should probably be explained, or be a link to an explanation somewhere.

Author: Hmm good point. Will expand.

Author: Done!

Delegation provides a way to "transfer authority without transferring cryptographic keys". As an authorization system, it is more interested in "what can be done" than a list of "who can do what". For more on how Delegation fits into UCAN, please refer to the [high level spec][UCAN].

# [UCAN Envelope] Configuration

## Type Tag

The UCAN envelope tag for UCAN Delegation MUST be set to `ucan/dlg@1.0.0-rc.1`.
 
## Delegation Payload

The Delegation payload MUST describe the authorization claims, who is involved, and its validity period.

| Field   | Type                                      | Required | Description                                                 |
|---------|-------------------------------------------|----------|-------------------------------------------------------------|
| `iss`   | `DID`                                     | Yes      | Issuer DID (sender)                                         |
| `aud`   | `DID`                                     | Yes      | Audience DID (receiver)                                     |

> | `sub`   | `DID`                                     | Yes      | Principal that the chain is about (the [Subject])           |

Reviewer: We discussed powerline concepts in https://github.com/ucan-wg/delegation/issues/11 and I believe came to an agreement that setting `sub: null` would imply re-delegation of everything that had been delegated to the issuer however spec does not seem to mention it and in fact table contradicts it. Can we please make sure that powerlines  are covered ?

```suggestion
| `sub`   | `DID \| null`                             | Yes      | Principal that the chain is about (the [Subject])           |
```

Author: Fixed ✅

| `meta`  | `{String : Any}`                          | No       | [Meta] (asserted, signed data) — is not delegated authority |
| `nbf`   | `Integer` (53-bits[^js-num-size])         | No       | "Not before" UTC Unix Timestamp in seconds (valid from)     |
| `exp`   | `Integer \| null` (53-bits[^js-num-size]) | Yes      | Expiration UTC Unix Timestamp in seconds (valid until)      |

# Capability

Capabilities are the semantically-relevant claims of a delegation. They MUST be presented as a map under the `cap` field as a map. This map is REQUIRED but MAY be empty. This MUST take the following form:

| Field  | Type      | Required | Description                                                                                      |
|--------|-----------|----------|--------------------------------------------------------------------------------------------------|
| `sub`  | `DID`     | Yes      | The [Subject] that this Capability is about                                                      |
| `cmd`  | `Command` | Yes      | The [Command] of this Capability                                                                 |

> | `pol`  | `Policy`  | Yes      | Additional constraints on eventual Invocation arguments, expressed in the [UCAN Policy Language] |

Author: Thanks!

Here is an illustrative example:

``` js
{
  // ...
  "sub": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp"
  "cmd": "/blog/post/create",
  "pol": [
    ["==", ".status", "draft"],

>     ["every", ".reviewer", ["match", ".email", "*@example.com"]],

Author: We had gone back and forth on `like` and `match` several times. The argument against `like` is that we don't support single character matching, so it could be confusing. But also, I bet that very few people will make the connection to SQL, and I don't really have strong opinions. I had `match` in the Rust code from way back when so went with that, but can switch to `like` since it sounds like you do care about it 👍

>     ["every", ".reviewer", ["match", ".email", "*@example.com"]],

Author: This needed changes in a bunch of places (hence not just merging the `suggestion`), but its' updated now ✅

    ["some", ".tags", 
      ["or",
        ["==", ".", "news"], 
        ["==", ".", "press"]]]
  ]
}
```

## Subject

The Subject MUST be the DID that initiated the delegation chain.

For example:

``` js
{
  "sub": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp",
  // ...
}
```

## Resource

Unlike Subjects and Commands, Resources are semantic rather than syntactic. The Resource is the "what" that a capability describes.

By default, the Resource of a capability is the Subject. This makes the delegation chain self-certifying.

``` js
{
  "sub": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp", // Subject
  // ...
}
```

In the case where access to an [external resource] is delegated, the Subject MUST own the relationship to the Resource. The Resource SHOULD be referenced by a `uri` key in the relevant [Conditions], except where it would be clearer to do otherwise. This MUST be defined by the Subject and understood by the executor.

``` js
{
  "sub": "did:key:z6MkiTBz1ymuepAQ4HEHYSF1H8quG5GLVVQR3djdX3mDooWp",
  "cmd": "/crud/create",
  "pol": [
    ["==", ".url", "https://example.com/blog/"], // Resource managed by the Subject
    // ...
  ],
  // ...
}
```

> 

Reviewer: This suggestion is pretty rough, but I think spec needs to cover commands and specifically aspects noted.

```suggestion

## Command

The [Command] MUST be a `/` delimited path describing set of commands delegated.

Delegation covers exact [Command] specified and all the commands described by a paths nested under that specified command. By definition `"/"` implies all of the command on a resource and SHOULD be used rarely and with a great care.

> ℹ️ Path based syntax is designed to support forward compatible protocol extensions - new capabilities could be introduced as command subpaths.
```

Author: > [!NOTE]
> I modified the syntax to use the built-in callouts, but essentially merged verbatim. Thanks!

# Policy Language

UCAN Delegation uses predicate logic statements extended with [jq]-style selectors as a policy language. Policies are syntactically driven, and MUST constrain the `args` field of an eventual [Invocation].

A Policy is always given as an array of predicates. This top-level array is implicitly treated as a logical `and`, where `args` MUST pass validation of every top-level predicate.

> Policies are structured as trees. With the exception of subtrees under `some`, `or`, and `not`, every leaf MUST evaluate to `true`. `some`, `or`, and `not` must 

Author: That looks like I fat fingered a `p` in doom-emacs 😅 Thanks

Glob patterns MUST only include one specicial character: `*` ("wildcard"). There is no single character matcher. As many `*`s as desired MAY be used. Non-wildcard `*`-literals MUST be escaped (`"\*"`). Attempting to match on a non-string MUST return false and MUST NOT throw an exception.

The wildcard represents zero-or-more characters. The following string literals MUST pass validation for the pattern `"Alice\*, Bob*, Carol.`:

* `"Alice*, Bob, Carol."`
* `"Alice*, Bob, Dan, Erin, Carol."`
* `"Alice*, Bob  , Carol."`
* `"Alice*, Bob*, Carol."`

The following MUST NOT pass validation for that same pattern:

* `"Alice*, Bob, Carol"` (missing the final `.`)
* `"Alice*, Bob*, Carol!"` (final `.` MUST NOT be treated as a wildcard)
* `"Alice, Bob, Carol."` (missing the `*` after `Alice`)
* `"Alice Cooper, Bob, Carol."` (the `*` after `Alice` is an escaped literal in the pattern)
* `" Alice*, Bob, Carol. "` (whitespace in the pattern is significant)

## Connectives

Connectives add context to their enclosed statement(s).

| Operator | Argument(s)   | Example                                    |
|----------|---------------|--------------------------------------------|

> | `not`    | `Statement`   | `["not", [">", ".a", 1]]`                  |

Author: > be able to enclose multiple statements

That may be confusing, no? Does it add an implicit `and`? Datomic likely does an implicit `or` due to how DNF works in datalog (which this is not)

> | `not`    | `Statement`   | `["not", [">", ".a", 1]]`                  |

Author: @Gozala

> Restricting negation to single cause is unnecessarily restricting and leaves undefined behavior when multiple clause may be present. 

Also I disagree about undefined behaviour when multiple clauses are present: that case is disallowed since multiple clauses after a `not` matches no forms.

| `and`    | `[Statement]` | `["and", [[">", ".a", 1], [">", ".b", 2]]` |

> | `or`     | `[Statement]` | `["or",  [[">", ".a", 1], [">", ".b", 2]]` | 

Author: @Gozala LOL yeah, I think I got stuck between the two options. The other way is to always force an array, but I'm happy either way. Will update per your comment.

```js
// Alternative
["and", [[">", ".a", 1], [">", ".b", 2]]] // I like this less
```

> | `or`     | `[Statement]` | `["or",  [[">", ".a", 1], [">", ".b", 2]]` | 

Author: (Will merge, but want to check the prose to make sure it's also clear about this case first)

> | `or`     | `[Statement]` | `["or",  [[">", ".a", 1], [">", ".b", 2]]` | 

Author: ...actually, if we want to follow `jq`, then we should always wrap this in an array. It's ugly, but that's how their `and` and `or` work:

<img width="705" alt="Screenshot 2024-08-05 at 16 57 59" src="https://github.com/user-attachments/assets/46736b6c-d012-446b-95d3-4283e76819ce">

`not` MUST invert the truth value of the inner statement. For example, if `["==", ".a", 1]` were false (`.a` is not 1), then `["not", ["==", ".a", 1]]` would be true.

`and` MUST take an arbitrarily long array of statements, and require that every inner statement be true. An empty array MUST be treated as vacuously true.

`or` MUST take an arbitrarily long array of statements, and require that at least one inner statement be true. An empty array MUST be treated as vacuously true.

## Quantification

> When a selector resolves to a collection (an array or map), quantifiers provide a way to extend `and` and `or` to their contents. Attempting to quantify over a non-collection MUST return false and MUST NOT throw an exception.

Reviewer: > Attempting to quantify over a non-collection MUST return false and MUST NOT throw an exception.

What is the rational for this decision ? I have being advocating that quantifying over non collection should behave as quantifying over single element collection because it is helpful in schema evolution when things 1:1 relations can change to 1:n relations.

In my experience with schema-on-read systems like datomic are a lot more flexible because they follow this approach.

I should clarify that my position implies that iteration over `null` should be equivalent to iteration over `[]` as opposed to `[null]`.

Reviewer: I do not believe iteration over map had being covered and if it is supported it needs to be clarified what `.` gets bound to for the maps.

Author: The rationale is that it breaks the policy assertion. Quantification implies that there is some collection. Much like using `?` elsewhere, you can always recover by wrapping in an `or` to say "or if it's not a collection, do this instead".

|---------------------------|----------------------------|-------------------------------------------------------------------------------------------------|
| Identity                  | `.`                        | Take the entire argument                                                                        |
| Dotted field name         | `.foo`, `.bar0_`           | Shorthand for selecting in a map by key (with exceptions, see below)                            |
| Unambiguous field name    | `["."]`, `["$_*"]`         | Select in a map by arbitrary key                                                                |
| Collection values         | `[]`                       | Expands out all of the children that match the remaining path FIXME double check with the code  |
| Collection index          | `[0]`, `[42]`              | The element at an index. On a map, this is decided by IPLD's key sort order. FIXME double check |
| Collection negative index | `[-1]`, `[-42]`            | The element by index from the end. `-1` is the index for the last element.                      |

> | Try                       | `.foo?`, `["nope"]?` | Returns `null` on what would otherwise fail                                                     |

Author: @Gozala sure 👍 I had thought we were kicking that down the road, but I won't block it. Adding...

> | Try                       | `.foo?`, `["nope"]?` | Returns `null` on what would otherwise fail                                                     |

Author: Fixed ✅

  
<td>

``` json
"Meeting Confirmation"
```

</td>
</tr>

<tr>
<td><pre>".cc"</pre></td>
<td>

``` json
["fraud@example.com"]
```

</td>
</tr>
    
<tr>
<td><pre>".to[1]"</pre></td>
<td>

``` json
"carol@not.example.com"
```

</td>
</tr>

<tr>
<td><pre>".to[-1]"</pre></td>
<td>

``` json
"dan@example.com"
```

</td>
</tr>

<tr>
<td><pre>".to[99]?"</pre></td>
<td>

``` json
null
```

</td>
</tr>
</tbody>
</table>

### Differences from jq

[jq] is a much larger language than UCAN's selectors. jq includes features like pipes, arithmatic, regexes, assignment, recursive descent, and so on which MUST NOT be supported in the UCAN Policy language.

jq produces streams of values, in contract to UCAN argument selectors which return an IPLD value. This introduces the primary difference between jq and UCAN argument selectors is how to treat output of the try (`?`) operator: UCAN's `try` selector operator MUST return `null` for the failure case.

## Validation

Validation involves substituting the values from the `args` field into the Policy, and evaluating the predicate. Since Policies are tree structured, selector substitution and predicate evaluation MAY proceed in any order.

> If a selector cannot be resolved (there is no value at that path), the associated statement MUST return false, and MUST NOT throw an exception. Note that for consistent semantics, selecting a missing keys on a map MUST return `null` (but nested selectors without a try MUST then fail the predicate).

Reviewer: This contradicts some of the prior agreements that made here https://github.com/ucan-wg/delegation/issues/5 specifically I made case that:

1. reference to missing key
2. out of bound index reference
3. Iteration over non collection

All must be errors which could be trapped using try `?` suffix to opt-in into return `null` behavior. My rational had been that optionality could be handled explicitly.

Author: Yeah I think is a case of me working from the Rust version where we had diverged while iterating quickly. You're right and the rationale makes sense to me; will fix 👍

Author: Ah, it's also how `jq` behaves:

```shell
> echo '{"a": 1}' | jq '.b.c.d.e.f'
null

> echo '[1]' | jq '.b'
jq: error (at <stdin>:1): Cannot index array with string "b"
```

When writing this, I had `jq` running on the side to check edge cases. Anywhere that we diverge from `jq` will need to be highlighted in big neon lights (or we need to do whatever they do since it's easier to manually test against / principle of least surprise)

Delegation provides a way to "transfer authority without transferring cryptographic keys". As an authorization system, it is more interested in "what can be done?" than a list of "who can do what?". For more on how Delegation fits into UCAN, please refer to the [high level spec][UCAN].

# Delegation (Envelope)

Delegations MUST include a signature that validates against the `iss` DID. A Delegation Payload on its own MUST NOT be considered a valid Delegation.

``` mermaid
flowchart TD
    subgraph Delegation
        SignatureBytes["Signature (raw bytes)"]
      
        subgraph SigPayload ["Signature Payload"]
            VarsigHeader["Varsig Header"]

            subgraph DelegationPayload ["Delegation Payload"]
                iss
                sub
                can
                args
                prf
                etc["..."]
            end
        end
    end
```

``` js
{
  "s": {"/": {"bytes": "7aEDQLYvb3lygk9yvAbk0OZD0q+iF9c3+wpZC4YlFThkiNShcVriobPFr/wl3akjM18VvIv/Zw2LtA4uUmB5m8PWEAU"}}
  "p": {
    "h": {"/": {"bytes": "NBIFEgEAcQ"}},

>     "ucan/d/1.0.0-rc.1": {

Reviewer 5: Is `aud` missing in the Delegation payload example?

      "iss": "did:plc:ewvi7nxzyoun6zhxrhs64oiz",
      "sub": "did:key:z6MkrZ1r5XBFZjBU34qyD8fueMbMRkKw17BZaq2ivKFjnz2z",
      "can": "crud/create",
      "args": {
        "uri": "https://example.com/blog/posts",
        "headers": {
          "content-type": "application/json"
        },
        "payload": {
          "title": "UCAN for Fun an Profit",
          "body": "UCAN is great!",
          "topics": ["authz", "journal"],
          "draft": true
        }
      },
      "nonce": {"/": {"bytes": "TWFueSBopvcs"}},
      "exp": 1697409438
    }
  }
}
```

> | Field | Type               | Required | Description                                           |

Reviewer 4: ```suggestion
| Field | JWT Property | Type               | Required | Description                                           |
```

Reviewer 2: ```suggestion
| Field | Type               | Required | Description                                           |
```

|-------|--------------------|----------|-------------------------------------------------------|
| `s`   | `Signature`        | Yes      | A signature by the Payload's `iss` over the `p` field |
| `p`   | `SignaturePayload` | Yes      | The content that was signed                           |

## Signature Payload

| Field               | Type                | Required | Description              |
|---------------------|---------------------|----------|--------------------------|

> | `h`                 | `VarsigHeader`      | Yes      | The Varsig header        |

Reviewer: I'm still not sure what goes here, I don't think it was covered anywhere in the spec

}

// Valid invocation
{
  "do": "msg/send",
  "args": {
    "from": "alice@example.com", // Matches above
    "to": ["bob@example.com", "carol@elsewhere.example.com"],
    "title": "Coffee",
    "body": "Still on for coffee" // Matches above
  },
  // ...
}
```

Any arguments MUST be taken verbatim and MUST NOT be further adjusted. For more flexible validation of Arguments, use [Conditions].

Note that this also applies to arrays and objects. For example, the `to` array in this example is considered to be exact, so the Invocation fails validation in this case:

``` js
// Delegation
{
  "can": "msg/send",
  "args": {
    "from": "alice@example.com",
    "to": ["bob@example.com"],
  },
  // ...
}

// INVALID Invocation
{
  "do": "msg/send",
  "args": {
    "from": "alice@example.com", // Matches above
    "to": ["bob@example.com", "carol@elsewhere.example.com"], // Does not match exactly
    "title": "Coffee",
    "body": "Still on for coffee"
  },
  // ...
}
```

The intended logic is expressible with [Conditions].

## Conditions

The `cond` field MUST contain any additional conditions. This concept is sometimes called a "caveat". Conditions constrain the capability in two ways:

> - Syntactic constraints on [Arguments] (length, regex, inclusion)

Reviewer 6: Reading the preview, and it looks like `[Arguments]` may need a link definition.

Condition semantics MUST be established by the Subject. They are openly extensible, but vocabularies may be reused across many Subjects. Conditions MUST be Understood by the [Executor] of the eventual [Invocation][UCAN Invocation]. Each Condition MUST be formatted as a map.

``` js
// Delegation
{
  "sub": "did:web:example.com",
  "can": "msg/send",
  "args": {
    "from": "alice@example.com"
  }
  "cond": [
     {                               // ┐
       "day": "monday",              // ├─ Condition
     },                              // ┘
     {                               // ┐
       "field": "to",                // ├─ Condition
       "includes": "bob@example.com" // │
     }                               // ┘
  ],
  // ...
}

// Valid Invocation, if Monday
{
  "do": "msg/send",
  "args": {
    "from": "alice@example.com",
    "to": ["bob@example.com", "carol@elsewhere.example.com"], // Matches criteria
    "title": "Coffee",
    "body": "Still on for coffee"
  },
  // ...
}
```
  
The above Delegation MUST be interpreted as "may send email from `alice@example.com` on Mondays as long as `bob@exmaple.com` is among the recipients"

> The `if` field MUST take the following shape: `[{}]`. The array represents a logical `all` (chained `AND`s). To represent logical `OR`, issue another delegation with that attenuation. For instance, the following represents `{a: 1, b:2} AND {c: 3} AND {d: 4}`:

Reviewer 6: ```suggestion
The `if` field MUST take the following shape: `[{}]`. The array represents a logical `all` (chained `AND`s). To represent logical `OR`, issue another delegation with that attenuation. For instance, the following represents `{a: 1, b: 2} AND {c: 3} AND {d: 4}`:
```

``` js
[
  {       // ┐

>     a: 1, // ├─ Confition ─┐

Reviewer 6: ```suggestion
    a: 1, // ├─ Condition ─┐
```

    rootCap -- references --> resource

    subgraph Delegations
        subgraph root [Root UCAN]
            subgraph rooting [Root Issuer]
                rootIss(iss: Alice)
                rootSub(sub: Alice)
            end

            rootCap("cap: (Storage, crud/*)")
            rootAud(aud: Bob)
        end

        subgraph del1 [Delegated UCAN]
            del1Iss(iss: Bob) --> rootAud
            del1Sub(sub: Alice)
            del1Aud(aud: Carol)
            del1Cap("cap: (Storage, crud/*)") --> rootCap

            del1Sub --> rootSub
        end

        subgraph del2 [Delegated UCAN]
            del2Iss(iss: Carol) --> del1Aud
            del2Sub(sub: Alice)
            del2Aud(aud: Dan)
            del2Cap("cap: (Storage, crud/*)") --> del1Cap

            del2Sub --> del1Sub
        end
    end

     subgraph inv [Invocation]
        invIss(iss: Dan)
        args("args: [Storage, crud/update, (key, value)]")
        invSub(sub: Alice)
        prf("proofs")
    end

    invIss --> del2Aud
    invoker --> invIss
    args --> del2Cap
    invSub --> del2Sub
    rootIss --> subject
    rootSub --> subject
    prf --> Delegations
```

### Recipient Validation

> An agent executing a capability MUST verify that the outermost `aud` field _matches its own DID._ The associated ability MUST NOT be performed if they do not match. Recipient validation is REQUIRED to prevent the misuse of UCANs in an unintended context.

Reviewer: Does this belong in the delegation spec ? Perhaps it should be moved to invocation one instead ? Or maybe it should be reframed so it's not about the execution but rather about received delegation ?

# [UCAN Envelope] Configuration
[UCAN Envelope Configuration]: #ucan-envelope-configuration
 
## Type Tag
[Type Tag]: #type-tag

The UCAN envelope tag for UCAN Delegation MUST be set to `ucan/dlg@1.0.0-rc.1`.
 
## Delegation Payload
[Delegation Payload]: #delegation-payload

The Delegation payload MUST describe the authorization claims, who is involved, and its validity period.

| Field   | Type                                      | Required | Description                                                 |
|---------|-------------------------------------------|----------|-------------------------------------------------------------|
| `iss`   | `DID`                                     | Yes      | Issuer DID (sender)                                         |
| `aud`   | `DID`                                     | Yes      | Audience DID (receiver)                                     |
| `sub`   | `DID \| null`                             | Yes      | Principal that the chain is about (the [Subject])           |

> | `cmd`   | `String`                                  | Yes      | The [Command] to eventually invoke                          |

Reviewer 7: Or maybe better, have it be a `[Capability]`, so that one policy remains attached to one command neatly.

> | `cmd`   | `String`                                  | Yes      | The [Command] to eventually invoke                          |

Author: We did essentially `[Capability]` previously, and it had some ugly tradeoffs in practice that we heard complaints about for _years_ until we removed. Some people liked it, but it made the token more complex, and saved way less space than you'd think. It also forces you to pass around (and leak!) a bunch of extra capabilities that don't apply to, it's way more complex to do revocations that target only one capability, etc.

The one thing that I do want to recover from `[Capability]` is fewer signatures (as always: crypto can be expensive). The solution there is to adopt batch signatures, which are flagged for a future version of Varsig. Basically build a Merkle tree of things to sign, sign the root, and include the Merkle proof to your token in the signature field.

  "sub": null, // 👈 ⚡ Powerline
  "cmd": "/crud/read",
  "pol": [],
  // ...
}
```

## Command
[Command]: #command

The [Command][High Level Command] MUST be a `/` delimited path describing set of commands delegated.
Delegation covers exact [Command] specified and all the commands described by a paths nested under that specified command. 

> [!NOTE]
> The command path syntax is designed to support forward compatible protocol extensions. Backwards-compatibl️️️️️️️️️️e capabilities MAY be introduced as command subpaths.

> [!WARNING]
> By definition `"/"` implies all of the commands available on a resource, and SHOULD be used with great care.

# Policy
[Policy]: #policy

UCAN Delegation uses predicate logic statements extended with [jq]-inspired selectors as a policy language. Policies are syntactically driven, and MUST constrain the `args` field of an eventual [Invocation].

A Policy is always given as an array of predicates. This top-level array is implicitly treated as a logical `and`, where `args` MUST pass validation of every top-level predicate.

Policies are structured as trees. With the exception of subtrees under `any`, `or`, and `not`, every leaf MUST evaluate to `true`. 

> A Policy is an array of statements. Every statement MUST take the form `[operator, selector, argument]` except for negation which MUST take the form `["not", statement]`.

Reviewer 7: Unless I'm missing something, `and` and `any` also have that 2-tuple form.

Reviewer 7: answering to myself: `and` and `or` can be *both* connectives (2-tuple) or quantifier (3-tuple).

I didn't see that one coming ;)

Reviewer 7: Hmm, those two are actually N-tuple, with N >= 2

Reviewer 7: No, they are 2-tuple, but the ABNF grammar is wrong

;; COMPARISONS

equality    = "[" DQUOTE "==" DQUOTE "," selector "," ipld   "]" ; Equality on IPLD literals
inequality  = "[" DQUOTE ">"  DQUOTE "," selector "," number "]" ; Numeric greater-than
            / "[" DQUOTE ">=" DQUOTE "," selector "," number "]" ; Numeric greater-than-or-equal
            / "[" DQUOTE "<"  DQUOTE "," selector "," number "]" ; Numeric lesser-than
            / "[" DQUOTE "<=" DQUOTE "," selector "," number "]" ; Numeric lesser-than-or-equal

wildcard    = "[" DQUOTE "like" DQUOTE "," selector "," pattern "]" ; String wildcard matching

;; SELECTORS

selector    = DQUOTE "." DQUOTE                     ; Identity
            / DQUOTE 1*(subselector *1("?")) DQUOTE ; Nested subselectors with possible optionals

subselector = "." CHAR string                           ; Dotted field selector
            / *1(".") "[\" DQUOTE string "\" DQUOTE "]" ; Explicit field selector
            / *1(".") "[" integer "]"                   ; Index selector
            / *1(".") "[]"                              ; Collection values // FIXME doble check code

;; SPECIAL LITERALS

pattern     = DQUOTE string DQUOTE ; Reminder: IPLD strings are UTF-8

> number      = integer / float

Reviewer 7: According to this line, selectors only operates on integers or floats, but below the spec also mention selecting on bytes (`Selecting on Bytes` section). Should bytes be included here?

Reviewer 7: nevermind, I got it wrong

|----------|--------------------------------|----------------------------------|
| `==`     | `Selector, IPLD`               | `["==", ".a", [1, 2, {"b": 3}]]` |
| `<`      | `Selector, (integer \| float)` | `["<",  ".a", 1]`                |
| `<=`     | `Selector, (integer \| float)` | `["<=", ".a", 1]`                |
| `>`      | `Selector, (integer \| float)` | `[">",  ".a", 1]`                |
| `>=`     | `Selector, (integer \| float)` | `[">=", ".a", 1]`                |

Literal equality (`==`) MUST match the resolved selecor to entire IPLD argument. This is a "deep comparison".

Numeric inequalities MUST be agnostic to numeric type. In other words, the decimal representation is considered equivalent to an integer (`1 == 1.0 == 1.00`). Attempting to compare a non-numeric type MUST return false and MUST NOT throw an exception.

## Glob Matching
[Glob Matching]: #glob-matching

| Operator | Argument(s)         | Example                               |
|----------|---------------------|---------------------------------------|
| `like`   | `Selector, Pattern` | `["like", ".email", "*@example.com"]` |

> Glob patterns MUST only include one specicial character: `*` ("wildcard"). There is no single character matcher. As many `*`s as desired MAY be used. Non-wildcard `*`-literals MUST be escaped (`"\*"`). Attempting to match on a non-string MUST return false and MUST NOT throw an exception.

Reviewer 7: is zero `*` allowed? Does that means exactly one `*` or at most one `*`?

Reviewer 8: +1

Reviewer 7: For the next person, I read that wrong:

> Glob patterns MUST only include one special character

means: the grammar only has one special character (`*`), but that character is allowed multiple times in the same pattern: `foo*bar*baz` is valid.

* `"Alice*, Bob, Carol."`
* `"Alice*, Bob, Dan, Erin, Carol."`
* `"Alice*, Bob  , Carol."`
* `"Alice*, Bob*, Carol."`

The following MUST NOT pass validation for that same pattern:

* `"Alice*, Bob, Carol"` (missing the final `.`)
* `"Alice*, Bob*, Carol!"` (final `.` MUST NOT be treated as a wildcard)
* `"Alice, Bob, Carol."` (missing the `*` after `Alice`)
* `"Alice Cooper, Bob, Carol."` (the `*` after `Alice` is an escaped literal in the pattern)
* `" Alice*, Bob, Carol. "` (whitespace in the pattern is significant)

## Connectives
[Connectives]: #connectives

Connectives add context to their enclosed statement(s).

| Operator | Argument(s)   | Example                                    |
|----------|---------------|--------------------------------------------|
| `and`    | `[Statement]` | `["and", [[">", ".a", 1], [">", ".b", 2]]` |
| `or`     | `[Statement]` | `["or",  [[">", ".a", 1], [">", ".b", 2]]` | 
| `not`    | `Statement`   | `["not", [">", ".a", 1]]`                  |

### And
[And]: #and

`and` MUST take an arbitrarily long array of statements, and require that every inner statement be true. An empty array MUST be treated as true.

```js
// Data

> { name: "Katie", age: 35, nationalities: ["Canadian", "South African"] }

Reviewer 7: ```suggestion
{ "name": "Katie", "age": 35, "nationalities": ["Canadian", "South African"] }
```

> ["and", []]

Reviewer 7: If I'm not mistaken, all those policy example should be wrapped with `[ <example goes here> ]`, otherwise there are not legal as per the syntax

Reviewer 7: I suppose those are statements, not policies, but that's a bit confusing.

// ⬆️  true

["and", [
  ["==", ".name", "Katie"], 
  [">=", ".age", 21]
]]
// ⬆️  true

["and", [
  ["==", ".name", "Katie"], 
  [">=", ".age", 21], 
  ["==", ".nationalities", ["American"]] // ️⬅️  false
]]
// ⬆️  false
```

### Or
[Or]: #or

`or` MUST take an arbitrarily long array of statements, and require that at least one inner statement be true. An empty array MUST be treated as true.

```js
// Data
{ name: "Katie", age: 35, nationalities: ["Canadian", "South African"] }

["or", []]
// ⬆️  true

["or", [
  ["==", ".name", "Katie"], // ⬅️  true
  [">", ".age", 45] 
]]
// ⬆️  true
```

### Not
[Not]: #not

`not` MUST invert the truth value of the inner statement. For example, if `["==", ".a", 1]` were false (`.a` is not 1), then `["not", ["==", ".a", 1]]` would be true.

```js
// Data
{ name: "Katie", nationalities: ["Canadian", "South African"] }

> ["not", 
>   ["and", [
>     ["==", ".name", "Katie"], 
>     ["==", ".nationalities", ["American"]] // ⬅️  false
> ]]

Reviewer 7: ```suggestion
["not", 
  ["and", [
    ["==", ".name", "Katie"], 
    ["==", ".nationalities", ["American"]] // ⬅️  false
  ]]
]
```

``` js
["all", ".newsletters",
  ["any", ".recipients", 
    ["==", ".email", "fraud@example.com"]]]
```

## Selectors
[Selectors]: #selectors

Selector syntax is closely based on [jq]'s "filters". They operate on an [Invocation]'s `args` object.

Selectors MUST only include the following features:

| Selector Name             | Examples                            | Notes                                                                                           |
|---------------------------|-------------------------------------|-------------------------------------------------------------------------------------------------|
| Identity                  | `.`                                 | Take the entire argument                                                                        |
| Dotted field name         | `.foo`, `.bar0_`                    | Shorthand for selecting in a map by key (with exceptions, see below)                            |
| Unambiguous field name    | `["."]`, `["$_*"]`                  | Select in a map by arbitrary key                                                                |

> | Collection values         | `[]`                                | Expands out all of the children that match the remaining path FIXME double check with the code  |

Reviewer 7: could you clarify what that does and/or provide an example?

> | Collection index          | `[0]`, `[42]`                       | The element at an index. On a map, this is decided by IPLD's key sort order. FIXME double check |

Reviewer 7: could you clarify what that does and/or provide an example?
